extensions/mssql/src/sparkFeature/dialog/dialogCommands.ts

```diff
@@ -49,31 +55,61 @@ export class OpenSparkJobSubmissionDialogCommand extends Command {
 
 	private async selectConnection(): Promise<SqlClusterConnection> {
 		let connectionList: azdata.connection.Connection[] = await this.apiWrapper.getActiveConnections();
-		let displayList: string[] = new Array();
 		let connectionMap: Map<string, azdata.connection.Connection> = new Map();
+		let selectedHost: string = undefined;
+		let showConnectionDialog = false;
+
+		// Filter invalid connections
+		if (connectionList && connectionList.length > 0) {
+			connectionList = connectionList.filter(conn => conn.providerName === constants.sqlProviderName && (<MssqlOptions><any>conn.options).server);
+		}
+		// Prompt choice if we have active connections
 		if (connectionList && connectionList.length > 0) {
-			connectionList.forEach(conn => {
-				if (conn.providerName === constants.sqlProviderName) {
-					displayList.push(conn.options.host);
-					connectionMap.set(conn.options.host, conn);
-				}
+			let selectConnectionMsg = localize('selectOtherServer', 'Select other SQL Server');
+			let displayList: string[] = [];
+			if (connectionList && connectionList.length > 0) {
+				connectionList.forEach(conn => {
+					let options: MssqlOptions = <any>conn.options;
+					displayList.push(options.server);
+					connectionMap.set(options.server, conn);
+				});
+			}
+			displayList.push(selectConnectionMsg);
+
+			selectedHost = await vscode.window.showQuickPick(displayList, {
+				placeHolder:
+					localize('sparkJobSubmission_PleaseSelectSqlWithCluster',
+						"Please select SQL Server with big data cluster.")
 			});
+			if (selectedHost === selectConnectionMsg) {
+				showConnectionDialog = true;
+				selectedHost = undefined;
+			}
+		} else {
+			showConnectionDialog = true;
+		}
+
+		// Show connection dialog if still don't have a server
+		if (showConnectionDialog) {
+			let connection = await azdata.connection.openConnectionDialog([constants.sqlProviderName]);
+			if (connection) {
+				let options: MssqlOptions = <any>connection.options;
+				connectionMap.set(options.server, connection);
+				selectedHost = options.server;
+				// Wait an appropriate timeout so that the serverInfo object can populate...
+				await timeout(150);
```

Comment:
do we really need this? is the openConnectionDialog returning before the connection.options are set?

Reply:
We do - `SqlClusterLookUp.getSqlClusterConnection` queries the `getServerInfo` function. If I don't do a timeout, it will be null and connect will fail. If I timeout, the call has time to set the connection in the `connectionManagementService._connectionStatusManager` object and works. Otherwise it isn't set yet, and I cannot figure out how to cleanly wait on / ensure that this is available.